ooniprobe/Model/NetworkMeasurement.mm

```diff
@@ -464,3 +464,221 @@ -(void)on_entry:(const char*)str{
+                    blocking = ANOMALY_RED;
+                }
+            }
+        }
+        
+        //TODO don't know if needed, maybe remove
+        //whatsapp_web_failure must be null, if != null (set test blocked)
+        if ([json objectForKey:@"whatsapp_web_failure"] != [NSNull null])
+            blocking = ANOMALY_RED;
+        
+        if (blocking > self.anomaly){
+            self.anomaly = blocking;
+            [TestStorage set_anomaly:self.test_id :blocking];
+        }
+    }
+}
+@end
+
+@implementation Telegram : NetworkMeasurement
+
+-(id) init {
+    self = [super init];
+    if (self) {
+        self.name = @"telegram";
+    }
+    return self;
+}
+
+-(void)run{
+    self.backgroundTask = [[UIApplication sharedApplication] beginBackgroundTaskWithExpirationHandler:^{
+        [[UIApplication sharedApplication] endBackgroundTask:self.backgroundTask];
+        self.backgroundTask = UIBackgroundTaskInvalid;
+    }];
+    [self run_test];
+}
+
+-(void) run_test {
+    mk::nettests::TelegramTest test;
+    [super init_common:test];
+    test.on_entry([self](std::string s) {
+        [self on_entry:s.c_str()];
+    });
+    test.start([self]() {
+        [self testEnded];
+    });
+}
+/*
+ for telegram: red if either "telegram_http_blocking" or "telegram_tcp_blocking" is true, OR if ""telegram_web_status" is "blocked"
+ the "*_failure" keys for telegram and whatsapp might indicate a test failure / anomaly
+ docs: https://github.com/TheTorProject/ooni-spec/blob/master/test-specs/ts-020-telegram.md#semantics
+ */
+-(void)on_entry:(const char*)str{
+    if (str != nil) {
+        if (!self.entry){
+            [TestStorage set_entry:self.test_id];
+            self.entry = TRUE;
+        }
+        NSError *error;
+        NSData *data = [[NSString stringWithUTF8String:str] dataUsingEncoding:NSUTF8StringEncoding];
+        NSDictionary *json = [NSJSONSerialization JSONObjectWithData:data options:0 error:&error];
+        int blocking = ANOMALY_GREEN;
+        
+        //telegram_http_blocking and telegram_tcp_blocking must be both false if null (set test anomaly), if at least one is true (set test blocked)
+        //telegram_web_status must be "ok", if null anomaly, if "blocked" (set test blocked)
+        //telegram_web_failure must be null, if != null (set test blocked)
+        NSArray *keys = [[NSArray alloc] initWithObjects:@"telegram_http_blocking", @"telegram_tcp_blocking", nil];
+        for (NSString *key in keys) {
+            if ([json objectForKey:key]){
+                if ([json objectForKey:key] == [NSNull null]) {
+                    if (blocking < ANOMALY_ORANGE)
+                        blocking = ANOMALY_ORANGE;
+                }
+                else if ([[json objectForKey:key] boolValue]) {
+                    blocking = ANOMALY_RED;
+                }
+            }
+        }
+        //TODO don't know if needed, maybe remove
```

Comment:
@joelanders review these two TODOS

Reply:
I think I would ignore "telegram_web_failure", but keep what you have around "telegram_web_status".